### PR TITLE
Add finding first 5 shows in a city skill

### LINKED ISSUE #6: Ask Artsy to recommend me a show in New York
From https://www.artsy.net/shows/new-york-ny-usa.

### PR DESCRIPTION
**Pull request by ashkan18**
# Feature
Add finding first 5 current shows by city. With this PR, Artsy's skill will be able to understand following commands:
```
Alexa, ask Artsy for shows in <city>.
Alexa, ask Artsy for shows around <city>.
Alexa, ask Artsy shows in <city>.
Alexa, ask Artsy shows around <city>.
```

Fixes #6 

# Changes
New `ShowsIntent` is added in `index.js` which is triggered by following utterances:
```
ShowsIntent	for shows in {CITY}
ShowsIntent	shows in {CITY}
ShowsIntent	for shows around {CITY}
ShowsIntent	shows around {CITY}
```
Once received these commands, it will:
- Geocode city name and find it's `lat`, `lon`
- Calls Artsy's `api/v1/shows` endpoint and passing `lat`, `lon` to get first five shows near this point and returns the results.

## Geocoding
Added dependency to [node-geocoder](https://github.com/nchaulet/node-geocoder) which is used to geocode city names to `lat`, `lon`. We are using `google` as geocoding provider, we need to make sure `GOOGLE_GEOCODING_API_KEY` is set in environment variables

## Other changes
Not sure if you like this format or not, but switched to JS [string interpolation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) in `index.js`. 
```js
res.say("Sorry, I couldn't find an artist " + value + ". Try again?"); // OLD
res.say(`Sorry, I couldn't find an artist ${value}. Try again?`); // NEW
```

### CODE REVIEW

**Comment by ashkan18** [comment]
Test is failing since we need to add `GOOGLE_GEOCODING_API_KEY` env variable, I don't have access to Travis project to add it there :(

File: DEPLOYMENT.md [hunk 1]
```diff
@@ -36,6 +36,12 @@ Install dependent packages and export schema and utterances.
 make build
 ```
 
+### Set Environment Variables
+Make sure `ARTSY_CLIENT_ID` and `ARTSY_CLIENT_SECRET` are set in your environment variables. These are used to access Artsy's API.
```
Comment (dblock) [inline DEPLOYMENT.md]: Could use a space after the `###` line.

File: functions/artsy/index.js [hunk 2]
```diff
@@ -4,6 +4,14 @@ var xapp = require('./models/xapp');
 var api = require('./models/api');
 var _ = require('underscore');
 var removeMd = require('remove-markdown');
+var NodeGeocoder = require('node-geocoder');
```
Comment (dblock) [inline functions/artsy/index.js]: Convention says it should be lowercase `nodeGeocoder`.

File: functions/artsy/index.js [hunk 3]
```diff
@@ -120,6 +128,62 @@ app.intent('AboutIntent', {
...
+    },
+    function(req, res) {
+        var city = req.slot('CITY');
+        console.log('app.ShowsIntent: ' + city);
+
+        geocoder.geocode(city)
+            .then(function(geoRes) {
+                geocodeResult = _.first(geoRes);
+                if (_.isEmpty(geoRes)) {
+                    res.say(`Sorry, I couldn't find ${city}. Try again?`);
```
Comment (dblock) [inline functions/artsy/index.js]: Add a log for the city that couldn't be found, I found it very useful to have that info when debugging and testing.

File: functions/artsy/index.js [hunk 4]
```diff
@@ -120,6 +128,62 @@ app.intent('AboutIntent', {
...
+                    res.shouldEndSession(false);
+                    res.send();
+                } else {
+                    api.instance().then(function(api) {
+                        api.findShows(geocodeResult.latitude, geocodeResult.longitude).then(function(results) {
+                            var intro = `Current exhibitions around ${city}`
```
Comment (dblock) [inline functions/artsy/index.js]: Add a log for the city that was found.

**Comment by dblock** [comment]
I made you admin on the repo (and gave Engineering r/w access) so you should be able to fix CI.

**Comment by dblock** [comment]
You should login to https://developer.amazon.com as it@artsymail.com to update the skill configuration and get the developer version working with the new intent. 

Login to https://console.aws.amazon.com for Lambda configuration, the only supported node version I think is 4.3, so I am not sure what that means for this PR. 

screen shot 2016-12-27 at 7 00 30 am

**Comment by dblock** [comment]
If you feel motivated, https://github.com/artsy/elderfield/issues/23 would make production deployment a lot easier, too.

**Comment by dblock** [comment]
Re-reading the code there's actually a bigger issue. The LITERAL slot [has been deprecated](https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/migrating-to-the-improved-built-in-and-custom-slot-types#replacing-literal-with-custom-slot-types), and needs to become a custom slot type.

**Comment by ashkan18** [comment]
Good point, didn't know about that. Looking deeper, actually there are some predefined Slot types we could use, I switched to [`AMAZON.US_CITY`](https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/built-in-intent-ref/slot-type-reference#us_city). We might be able to change our artists custom slot type to also follow some of the predefined ones. Maybe [`AMAZON.ARTISTS`](https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/built-in-intent-ref/slot-type-reference#artist) although their sample list is pretty irrelevant.

**Comment by ashkan18** [comment]
Once https://github.com/artsy/elderfield/pull/36 is merged, I merge master to this branch to resolve conflicts and switch to new `api_base_url` changes and then this PR should be good to go.

Im still struggling with deploy process and whats the best way to do code upgrades.

**Comment by dblock** [comment]
You should be able to `apex deploy` and have it work on an actual echo or on https://echosim.io. The development version of the app is available automatically on any echo or echoism under the it@artsymail.com account (I have a dot on my desk that you should grab if you haven't yet).

Assuming you can test this code on a real device this can be merged.

**Comment by dblock** [comment]
Excellent work, merged!

**Pull request merged**